temporalio/worker/_tuning.py

```diff
@@ -70,7 +74,225 @@ class ResourceBasedSlotSupplier:
+        released. The permit will, however, be forgotten and python will garbage collect it. So if you use the same slot
+        supplier over the lifetime of more than one worker and it is critically important for you to clean up some
+        resources associated all permits you construct, then consider using a finalizer on your returned permits.
+
+        Args:
+            ctx: The context for slot reservation.
+
+        Returns:
+            A permit to use the slot which may be populated with your own data.
+        """
+        ...
+
+    @abstractmethod
+    def try_reserve_slot(self, ctx: SlotReserveContext) -> Optional[SlotPermit]:
+        """This function is called when trying to reserve slots for "eager" workflow and activity tasks.
+        Eager tasks are those which are returned as a result of completing a workflow task, rather than
+        from polling. Your implementation must not block, and if a slot is available, return a permit
+        to use that slot.
+
+        Args:
+            ctx: The context for slot reservation.
+
+        Returns:
+            Maybe a permit to use the slot which may be populated with your own data.
+        """
+        ...
+
+    @abstractmethod
+    def mark_slot_used(self, ctx: SlotMarkUsedContext) -> None:
+        """This function is called once a slot is actually being used to process some task, which may be
+        some time after the slot was reserved originally. For example, if there is no work for a
+        worker, a number of slots equal to the number of active pollers may already be reserved, but
+        none of them are being used yet. This call should be non-blocking.
+
+        Args:
+            ctx: The context for marking a slot as used.
+        """
+        ...
+
+    @abstractmethod
+    def release_slot(self, ctx: SlotReleaseContext) -> None:
+        """This function is called once a permit is no longer needed. This could be because the task has
+        finished, whether successfully or not, or because the slot was no longer needed (ex: the number
+        of active pollers decreased). This call should be non-blocking.
+
+        Args:
+            ctx: The context for releasing a slot.
+        """
+        ...
+
+
+SlotSupplier: TypeAlias = Union[
+    FixedSizeSlotSupplier, ResourceBasedSlotSupplier, CustomSlotSupplier
+]
+
+
+class _BridgeSlotSupplierWrapper:
+    def __init__(self, supplier: CustomSlotSupplier):
+        self._supplier = supplier
+
+    async def reserve_slot(
+        self, ctx: SlotReserveContext, reserve_cb: Callable[[Any], None]
+    ) -> SlotPermit:
+        try:
+            reserve_fut = asyncio.create_task(self._supplier.reserve_slot(ctx))
+            reserve_cb(reserve_fut)
+            return await reserve_fut
+        except asyncio.CancelledError:
+            raise
+        except Exception:
+            logger.warning(
```

Comment:
If you'd rather, instead of logging and having a global logger, can call https://docs.python.org/3/library/warnings.html#warnings.warn instead. We do this in other places for should-not-happen things without relying on them setting up logging.

Reply:
Decided not to do this since formatting for the user thrown exception is not nearly as nice w/ the stack trace not being shown whereas logging does